### ISSUE TITLE
how to link multispans to the same traceId

### ISSUE THREAD

**Issue by Karlzzb** [apache/skywalking#3738]
Please answer these questions before submitting your issue.

- Why do you submit this issue?
- [ ] Question or discussion
- [*] Bug
- [ ] Requirement
- [ ] Feature or performance improvement

___
### Question
- What do you want to know?
I need to know how to link two spans. 
___
### Bug
- Which version of SkyWalking, OS and JRE?
6.4.0
- Which company or project?
zcsmart
The inner framework in our company. https://github.com/karl-free-legion/zcs-legion-cluster.git
- What happen?
If possible, provide a way for reproducing the error. e.g. demo application, component version.

I need to know how to link two spans. I try two methods, neither of which is working.

First : I tried the way attach in my message, codes as below

```java       
CarrierItem next = contextCarrier.items();
        while (next.hasNext()) {
            next = next.next();
            String contextValue = header.getContextCarrierOrDefault(next.getHeadKey(),"");
            if (!StringUtil.isEmpty(contextValue)) {
                System.out.println("receive message ["+header.getTrackId()+"] contextCarrier got key="+next.getHeadKey()+",value="+contextValue);
                next.setHeadValue(contextValue);
            }
        }
```

Second: I also use the traceId setting way, such as
``` java       
List<DistributedTraceId> distributedTraceIds = new ArrayList();
        distributedTraceIds.add(new PropagatedTraceId(header.getTrackId()));
        contextCarrier.setDistributedTraceIds(distributedTraceIds);
        AbstractSpan span = ContextManager.createEntrySpan(header.getTag(), contextCarrier);
//        span.setPeer(header.getModuleSource().getGroupId());
        new StringTag(Constans.CARRIER_LEGION_TRACK_ID_KEY).set(span,header.getTrackId());
        span.setOperationName("handler Message【"+header.getTrackId()+"】");
        Tags.URL.set(span,header.getUri());
        SpanLayer.asRPCFramework(span);
        span.setComponent(ComponentsDefine.GRPC);
```
**The result is** 
端点:sent out Message【65.96.15723518143630001】
traceId: 65.96.15723518143630001
跨度类型:Exit
组件:
Peer:g001
失败:false
url:M://g001
legion_track_id:65.96.15723518143630001

跨度信息
标记.
端点:handler Message【65.96.15723518143630001】
traceId: 64.91.15723518147410001
跨度类型:Entry
组件:GRPC
Peer:No Peer
失败:false
legion_track_id:65.96.15723518143630001
url:M://g001

___
### Requirement or improvement
- Please describe about your requirements or improvement suggestions.

**Comment by wu-sheng** [comment]
Your code format is not right, could you adjust that in markdown? It is hard to follow.

**Comment by Karlzzb** [comment]
Thanks for responding


1. The code for sending message

```java
        X.XMessage rawMessage = (X.XMessage) allArguments[0];
        X.XHeader header = rawMessage.getHeader();
        ContextCarrier contextCarrier  = new ContextCarrier();
        RequestURI request = RequestURI.create(header.getUri());
        AbstractSpan span = ContextManager.createExitSpan(header.getTag(), contextCarrier, request.getGroupId());
        Tags.URL.set(span,header.getUri());
        X.XHeader.Builder newHeader = header.toBuilder();
        newHeader.setTrackId(contextCarrier.getDistributedTraceId().encode());
        span.setOperationName("sent out Message【"+newHeader.getTrackId()+"】");
        span.setComponent(ComponentsDefine.UNDERTOW);
        new StringTag(Constans.CARRIER_LEGION_TRACK_ID_KEY).set(span,newHeader.getTrackId());
        CarrierItem next = contextCarrier.items();
        while (next.hasNext()) {
            next = next.next();
            System.out.println("message=["+header.getTrackId()+"]  traceId=["+contextCarrier.getDistributedTraceId().encode()+"]  sent contextCarrier key="+next.getHeadKey()+",value="+next.getHeadValue());
            newHeader.putContextCarrier(next.getHeadKey(), next.getHeadValue());
        }
        rawMessage = rawMessage.toBuilder().setHeader(newHeader).build();
        allArguments[0] = rawMessage;
```

2. The code for handler receiving message

```java
        X.XMessage rawMessage = (X.XMessage) allArguments[0];
        X.XHeader header = rawMessage.getHeader();
        RequestURI request = RequestURI.create(header.getUri());
        System.out.println("X contextCarrier ="+header.getContextCarrierCount());
        for(Map.Entry e : header.getContextCarrierMap().entrySet()) {
            System.out.println("X contextCarrier key="+e.getKey()+",value="+e.getValue());
        }
        ContextCarrier contextCarrier  = new ContextCarrier();
        CarrierItem next = contextCarrier.items();
        while (next.hasNext()) {
            next = next.next();
            String contextValue = header.getContextCarrierOrDefault(next.getHeadKey(),"");
            if (!StringUtil.isEmpty(contextValue)) {
                System.out.println("receive message ["+header.getTrackId()+"] contextCarrier got key="+next.getHeadKey()+",value="+contextValue);
                next.setHeadValue(contextValue);
            }
        }
        List<DistributedTraceId> distributedTraceIds = new ArrayList();
        distributedTraceIds.add(new PropagatedTraceId(header.getTrackId()));
        contextCarrier.setDistributedTraceIds(distributedTraceIds);
        AbstractSpan span = ContextManager.createEntrySpan(header.getTag(), contextCarrier);
        new StringTag(Constans.CARRIER_LEGION_TRACK_ID_KEY).set(span,header.getTrackId());
        span.setOperationName("handler Message【"+header.getTrackId()+"】");
        span.setPeer(header.getModuleSource().getGroupId());
        Tags.URL.set(span,header.getUri());
        span.setComponent(ComponentsDefine.UNDERTOW);
```

**Comment by wu-sheng** [comment]
You are not following SkyWalking core API. Why you change the `contextCarrier` by your self? Once you changed, I don't know what you mean.

**Comment by Karlzzb** [comment]
I follow the example code  just like below, but it is not working.  So I just tried to set the traceId by myself is also not helping.
```java
contextCarrier.setDistributedTraceIds(distributedTraceIds);
```

1、Client side steps by Apache HTTPComponent client plugin
```java
            span = ContextManager.createExitSpan("/span/operation/name", contextCarrier, "ip:port");
            CarrierItem next = contextCarrier.items();
            while (next.hasNext()) {
                next = next.next();
                httpRequest.setHeader(next.getHeadKey(), next.getHeadValue());
            }
```
2、Server side steps by Tomcat 7 server plugin
```java
            ContextCarrier contextCarrier = new ContextCarrier();
            CarrierItem next = contextCarrier.items();
            while (next.hasNext()) {
                next = next.next();
                next.setHeadValue(request.getHeader(next.getHeadKey()));
            }

            span = ContextManager.createEntrySpan(“/span/operation/name”, contextCarrier);
```

**Comment by wu-sheng** [comment]
You are doing this
```
  distributedTraceIds.add(new PropagatedTraceId(header.getTrackId()));
  contextCarrier.setDistributedTraceIds(distributedTraceIds);
```
Can't understand why.

**Comment by Karlzzb** [comment]
I have client and service endpoint, need to bind two spans with the same traceId. I use the way transfer CarrierItem in my Message Header. However, it is not working. 

I might not have deep understanding. Could you guide me abouth how to link two spans with the same traceId? really appreciate.

**Comment by wu-sheng** [comment]
Do you read the existing MQ codes? ContextCarrier has APIs to deal with it already. Clearly, you are using the incorrect way.

**Comment by Karlzzb** [comment]
It is working! Thanks a lot!
